src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -218,7 +167,128 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
+            }
+
+            // Process the export if:
+            // 1. There is continuation token, which means there is more resource to be exported.
+            // 2. There is no continuation token but the page is 0, which means it's the initial export.
+            while (progress.ContinuationToken != null || progress.Page == 0)
+            {
+                SearchResult searchResult = null;
+
+                // Search and process the results.
+                using (IScoped<ISearchService> searchService = _searchServiceFactory())
+                {
+                    if (compartmentSearch)
+                    {
+                        searchResult = await searchService.Value.SearchCompartmentAsync(
+                            KnownResourceTypes.Patient,
+                            progress.TriggeringResourceId,
+                            null,
+                            queryParametersList,
+                            cancellationToken);
+                    }
+                    else
+                    {
+                        switch (_exportJobRecord.ExportType)
+                        {
+                            case ExportJobType.All:
+                                searchResult = await searchService.Value.SearchAsync(
+                                    null,
+                                    queryParametersList,
+                                    cancellationToken);
+                                break;
+                            case ExportJobType.Patient:
+                                searchResult = await searchService.Value.SearchAsync(
+                                    KnownResourceTypes.Patient,
+                                    queryParametersList,
+                                    cancellationToken);
+                                break;
+                        }
+                    }
+                }
+
+                if (_exportJobRecord.ExportType == ExportJobType.Patient && !compartmentSearch)
+                {
+                    uint resultIndex = 0;
+                    foreach (SearchResultEntry result in searchResult.Results)
+                    {
+                        // If a job is resumed in the middle of processing patient compartment resources it will skip patients it has already exported compartment information for.
+                        if (progress.SubSearch != null && result.Resource.ResourceId != progress.SubSearch.TriggeringResourceId)
+                        {
+                            continue;
+                        }
+
+                        if (progress.SubSearch == null)
+                        {
+                            progress.NewSubSearch(result.Resource.ResourceId);
+                        }
+
+                        await RunExportSearch(exportJobConfiguration, progress.SubSearch, sharedQueryParametersList, cancellationToken, currentBatchId + ":" + resultIndex.ToString("d6"), true);
```

Comment:
nit: I feel like using the current method recursively for the compartment search makes the code harder to read/debug. There are similarities and subtle differences in behavior. Would it be better to have something like `RunExportCompartmentSearch` that that deals only with compartment search scenario? That will also reduce the number of if conditions in the current method. Some of the common code could be extracted into helper methods if required.

Reply:
This is mostly my preference showing, I really like recursive methods :) 
I agree though that this isn't the best place for it, I'll refactor.